### PR TITLE
add support for built-in unit tls

### PR DESCRIPTION
**Pull request by quixoten**
## New Behavior

This copies some behavior from the [unit docker-entrypoint.sh](https://github.com/nginx/unit/blob/master/pkg/docker/docker-entrypoint.sh) so that it's easy to configure TLS locally. Specifically, it loops through all the `*.pem` and `*.json` in the `/etc/unit/` directory and loads them into [nginx/unit](https://unit.nginx.org/).

## Contrast to Current Behavior

Currently, only a single json config file can be loaded into unit, which makes using unit's built-in TLS impossible without building a custom image or overriding the existing load script with a volume mount.

## Discussion: Benefits and Drawbacks

The main benefit is making it easy to add TLS without involving external containers running caddy or some other tool to terminate the TLS connection. This change was written to be completely backwards compatible, and in most cases will be transparent to any users of the image. The only exception is users that are explicitly setting `UNIT_CONFIG` in the containers environment. They will see a deprecation warning, but the image will continue to work for them:

```
⚠️ The UNIT_CONFIG environment variable is deprecated. All *.pem and *.json files in /etc/unit will be loaded automatically when UNIT_CONFIG is undefined.
⚙️ Applying configuration from UNIT_CONFIG environment variable: /etc/unit/nginx-unit.json
```

## Changes to the Wiki

The wiki can be modified to add an example of using the built-in unit TLS. The caddy example does not need to be removed, since it'll still work if users want to use that method instead.

### Example Using unit's built-in TLS termination

To use unit's built-in TLS termination, mount a local directory into your netbox container at `/etc/unit` that contains the following two files: `nginx-unit.json` and `local-bundle.pem`.

You can create `local-bundle.pem` by installing [mkcert](https://github.com/FiloSottile/mkcert) and running the following commands:

```
mkcert netbox.local.gd localhost 127.0.0.1 ::1
cat netbox.local.gd+?.pem $(mkcert -CAROOT)/rootCA.pem netbox.local.gd+?-key.pem > local-bundle.pem
```

<details>
<summary>You're `nginx-unit.json` should look like this:</summary>

```
{
  "listeners": {
    "*:8080": {
      "pass": "routes",
      "tls": {
        "certificate": "local-bundle"
      }
    }
  },

  "routes": [
    {
      "match": {
        "uri": "/static/*"
      },
      "action": {
        "share": "/opt/netbox/netbox${uri}"
      }
    },

    {
      "action": {
        "pass": "applications/netbox"
      }
    }
  ],

  "applications": {
    "netbox": {
      "type": "python 3",
      "path": "/opt/netbox/netbox/",
      "module": "netbox.wsgi",
      "home": "/opt/netbox/venv",
      "processes": {
        "max": 4,
        "spare": 1,
        "idle_timeout": 120
      }
    }
  },

  "access_log": "/dev/stdout"
}
```
</details>

This is identical to the default `nginx-unit.json` except for the addition of the `tls` block starting on line 5. Note that `local-bundle` in the `tls` block matches the name of the `local-bundle.pem` file created earlier (with the `.pem` suffix removed). If you want to use a different name for your certificate bundle file, make sure the change is also reflected in the json configuration.

## Proposed Release Note Entry

Added support for terminating TLS connections internally. For an example, visit the [TLS section of the wiki](https://github.com/netbox-community/netbox-docker/wiki/TLS)

## Double Check

* [x] I have read the comments and followed the PR template.
* [x] I have explained my PR according to the information in the comments.
* [x] My PR targets the `develop` branch.

### CODE REVIEW

File: docker/launch-netbox.sh [hunk 1]
```diff
@@ -1,8 +1,33 @@
...
 
+put_config() {
+  RET=$(
+    curl \
+    --silent \
+    --write-out '%{http_code}' \
+    --request PUT \
+    --data-binary "@$1" \
+    --unix-socket $UNIT_SOCKET \
+    http://localhost/$2
```
Comment (cimnine) [inline docker/launch-netbox.sh:13]:
```suggestion
    "http://localhost/$2"
```

File: docker/launch-netbox.sh [hunk 2]
```diff
@@ -1,8 +1,33 @@
...
+  )
+  RET_BODY=${RET::-3}
+  RET_STATUS=$(echo $RET | tail -c 4)
+
+  if [ "$RET_STATUS" -ne "200" ]; then
+      echo "⚠️ Error: Failed to load configuration from $1"
```
Comment (cimnine) [inline docker/launch-netbox.sh:19]:
```suggestion
      echo "⚠️ Error: Failed to load configuration from '$1' to 'http://localhost/$2' (via '$UNIT_SOCKET')"
```

File: docker/launch-netbox.sh [hunk 3]
```diff
@@ -1,8 +1,33 @@
...
+      ( echo "HTTP response status code is '$RET_STATUS'"
+        echo "$RET_BODY"
+      ) | sed 's/^/  /'
```
Comment (cimnine) [inline docker/launch-netbox.sh:22]: What is the need for / intention of `sed` here?

File: docker/launch-netbox.sh [hunk 4]
```diff
@@ -22,27 +47,37 @@ load_configuration() {
...
+  if [[ -n "$UNIT_CONFIG" ]] && [[ -s "$UNIT_CONFIG" ]]; then
+    echo "⚠️ The UNIT_CONFIG environment variable is deprecated. All *.pem and *.json files in /etc/unit will be loaded automatically when UNIT_CONFIG is undefined."
+    echo "⚙️ Applying configuration from UNIT_CONFIG environment variable: $UNIT_CONFIG"
+    put_config $UNIT_CONFIG "config" || return 1
+  else
+    echo "🔍 Looking for certificate bundles in /etc/unit/..."
+    for f in $(find /etc/unit/ -type f -name "*.pem"); do
+      echo "⚙️ Uploading certificates bundle: $f"
+      put_config $f "certificates/$(basename $f .pem)" || return 1
+    done
```
Comment (cimnine) [inline docker/launch-netbox.sh:59]:
```suggestion
    for pem_file in /etc/unit/*.pem; do
      echo "⚙️ Uploading certificates bundle '$pem_file'"
      put_config $f "certificates/${pem_file%.json}" || return 1
    done
```

I believe the invocations of `find` and `baseline` should not be necessary.

File: docker/launch-netbox.sh [hunk 5]
```diff
@@ -22,27 +47,37 @@ load_configuration() {
...
-  )
-  if [ "$RESP_CODE" != "200" ]; then
-    echo "⚠️ Could no load Unit configuration"
-    kill "$(cat /opt/unit/unit.pid)"
-    return 1
+    echo "🔍 Looking for configuration snippets in /etc/unit/..."
+    for f in $(find /etc/unit/ -type f -name "*.json"); do
+      echo "⚙️ Applying configuration $f";
+      put_config $f "config" || return 1
+    done
```
Comment (cimnine) [inline docker/launch-netbox.sh:65]:
```suggestion
    for config_file in /etc/unit/*.json; do
      echo "⚙️ Applying configuration '$config_file'";
      put_config "$config_file" "config" || return 1
    done
```

File: docker/launch-netbox.sh [hunk 6]
```diff
@@ -22,27 +47,37 @@ load_configuration() {
...
+
+    # warn on filetypes we don't know what to do with
+    for f in $(find /etc/unit/ -type f -not -name "*.json" -not -name "*.pem"); do
+      echo "↩️ Ignoring $f";
+    done
```
Comment (cimnine) [inline docker/launch-netbox.sh:70]:
```suggestion
    shopt -s extglob
    for other_file in !(/etc/unit/*.json|/etc/unit/*.pem); do
      echo "↩️ Ignoring '$other_file'";
    done
    shopt -u extglob
```

File: docker/launch-netbox.sh [hunk 7]
```diff
@@ -22,27 +47,37 @@ load_configuration() {
...
   fi
 
   echo "✅ Unit configuration loaded successfully"
 }
 
+if [ -n "$(ls -A /opt/unit/state)" ]; then
+  echo "💣 Clearing previous unit state from /opt/unit/state"
+  rm -r /opt/unit/state/*
+fi
+
```
Comment (cimnine) [inline docker/launch-netbox.sh:80]: Why is that necessary?

File: docker/launch-netbox.sh (hunk 1 above)
Comment (cimnine) [inline docker/launch-netbox.sh:13]: I think I'd change the order of the arguments, so that the 'more' static parts are first. What I mean:

```suggestion
    --data-binary "@${2}" \
    --unix-socket $UNIT_SOCKET \
    "http://localhost/${1}"
```

The invocation would then be `put_config "config" "file.json"` and `put_config "certificate/bla" "bla.pem"`.
But I don't see any objective argument for one or the other, so if you like your argument order that's good with me.

File: docker/launch-netbox.sh [hunk 8]
```diff
@@ -1,8 +1,33 @@
...
+    curl \
+    --silent \
+    --write-out '%{http_code}' \
+    --request PUT \
+    --data-binary "@$1" \
+    --unix-socket $UNIT_SOCKET \
+    http://localhost/$2
+  )
+  RET_BODY=${RET::-3}
+  RET_STATUS=$(echo $RET | tail -c 4)
```
Comment (cimnine) [inline docker/launch-netbox.sh:16]:
```suggestion
  RET_BODY=${RET:0:-3}
  RET_STATUS=${RET: -3:3}     # Keep the space before the minus!
```

I believe an invocation of a subshell that runs `echo` and `tail` should not be necessary.

File: docker/launch-netbox.sh [hunk 9]
```diff
@@ -1,8 +1,33 @@
...
+
+  if [ "$RET_STATUS" -ne "200" ]; then
+      echo "⚠️ Error: Failed to load configuration from $1"
+      ( echo "HTTP response status code is '$RET_STATUS'"
+        echo "$RET_BODY"
+      ) | sed 's/^/  /'
+
+      kill "$(cat /opt/unit/unit.pid)"
```
Comment (cimnine) [inline docker/launch-netbox.sh:24]:
```suggestion
      kill < /opt/unit/unit.pid
```

I believe the invocation of the subshell and `cat` is not necessary.

File: docker/launch-netbox.sh [hunk 10]
```diff
@@ -22,27 +47,37 @@ load_configuration() {
   # this curl call will get a reply once unit is fully launched
   curl --silent --output /dev/null --request GET --unix-socket $UNIT_SOCKET http://localhost/
 
-  echo "⚙️ Applying configuration from $UNIT_CONFIG"
+  if [[ -n "$UNIT_CONFIG" ]] && [[ -s "$UNIT_CONFIG" ]]; then
+    echo "⚠️ The UNIT_CONFIG environment variable is deprecated. All *.pem and *.json files in /etc/unit will be loaded automatically when UNIT_CONFIG is undefined."
+    echo "⚙️ Applying configuration from UNIT_CONFIG environment variable: $UNIT_CONFIG"
```
Comment (cimnine) [inline docker/launch-netbox.sh:52]:
```suggestion
    echo "⚙️ Applying configuration from UNIT_CONFIG environment variable: '$UNIT_CONFIG'"
```

I dislike that the variable would become deprecated. I suggest to complement the variable with a new `$UNIT_PEM_FILE` variable that loads just that one certificate bundle – if defined.

File: docker/launch-netbox.sh [hunk 11]
```diff
@@ -22,27 +47,37 @@ load_configuration() {
...
+    put_config $UNIT_CONFIG "config" || return 1
+  else
+    echo "🔍 Looking for certificate bundles in /etc/unit/..."
```
Comment (cimnine) [inline docker/launch-netbox.sh:55]: I dislike that the folder is now hardcoded to `/etc/unit`.

```suggestion
    UNIT_CONFIG_DIR=${UNIT_CONFIG_DIR-/etc/unit/}
    echo "🔍 Looking for certificate bundles in '${UNIT_CONFIG_DIR}'"
```

Instead of `/etc/unit/` you'd just use the `UNIT_CONFIG_DIR` variable instead.

File: docker/launch-netbox.sh [hunk 12]
```diff
@@ -22,27 +47,37 @@ load_configuration() {
   # this curl call will get a reply once unit is fully launched
   curl --silent --output /dev/null --request GET --unix-socket $UNIT_SOCKET http://localhost/
 
-  echo "⚙️ Applying configuration from $UNIT_CONFIG"
+  if [[ -n "$UNIT_CONFIG" ]] && [[ -s "$UNIT_CONFIG" ]]; then
```
Comment (cimnine) [inline docker/launch-netbox.sh:50]:
```suggestion
  if [[ -n "$UNIT_CONFIG" ]] && [[ -s "$UNIT_CONFIG" ]]; then
```

Why does it need two checks? Are the double-brackets actually necessary?

**Review by cimnine — COMMENTED** [review]
I've added some suggestions to make the code more robust and to use more of bash's built-in functionality. Also, please run the script with your changes through [`shellcheck`](https://www.shellcheck.net/) and apply it's suggestions.

I like how the technical side of the TLS configuration challenge (especially loading the `pem` files) was solved in the PR. I also believe the approach could be improved to retain backwards compatibility with ease.

In general, I'm still against terminating TLS in the container. For the technical challenge, this solution seems like a good trade-off. But that was never the reason we don't recommend doing it that way. It's the fear of users who break the default configuration and submit bugs that are impossible to debug for us.

Especially if we make adjustments to the configuration in the future. Wouldn't that mean all users who configured TLS directly in unit will have to make adjustments to their independently managed unit configuration?

For a lot of users it's weird enough that they have to have a `docker-compose.yaml` file that matches the container version. If we offer them an easy way to configure TLS in the container, and then ask them to keep their independent configuration in sync with ours, then I believe we ask too much.

File: docker/launch-netbox.sh [hunk 13]
```diff
@@ -1,8 +1,33 @@
...
 UNIT_SOCKET="/opt/unit/unit.sock"
 
+put_config() {
+  RET=$(
+    curl \
+    --silent \
+    --write-out '%{http_code}' \
+    --request PUT \
+    --data-binary "@$1" \
+    --unix-socket $UNIT_SOCKET \
```
Comment (cimnine) [inline docker/launch-netbox.sh:12]:
```suggestion
    --unix-socket "$UNIT_SOCKET" \
```

**Comment by ryanmerolle** [comment]
pem is good, but some users leverage only a crt and key file also.
